h2o-mapreduce-generic/src/main/java/water/hadoop/FSCloudingEventSource.java

```diff
@@ -0,0 +1,150 @@
+package water.hadoop;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+class FSCloudingEventSource {
+
+  private final FileSystem _fs;
+  private final Path _path;
+
+  private final Set<String> _discovered = new HashSet<>();
+  
+  FSCloudingEventSource(Configuration conf, String cloudingDir) throws IOException  {
+    _fs = FileSystem.get(conf);
+    _path = makeQualifiedPath(conf, new Path(cloudingDir));
+  }
+
+  boolean isEmpty() throws IOException {
+    return !_fs.exists(_path);
+  }
+
+  Path getPath() {
+    return _path;
+  }
+
+  private static Path makeQualifiedPath(Configuration conf, Path cloudingPath) throws IOException {
+    return cloudingPath.getFileSystem(conf).makeQualified(cloudingPath);
+  }
+  
+  Iterable<CloudingEvent> fetchNewEvents() {
+    try {
+      if (isEmpty())
+        return Collections.emptyList();
+      List<CloudingEvent> newEvents = Arrays.stream(_fs
+              .listStatus(_path, path -> !_discovered.contains(path.getName())))
+              .map(it -> CloudingEvent.fromFileStatus(_fs, it))
+              .filter(Objects::nonNull)
+              .sorted(Collections.reverseOrder())
+              .collect(Collectors.toList());
+      newEvents.forEach(it -> _discovered.add(it._path.getName()));
+      return newEvents;
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to read events", e);
+    }
+  }
+
+}
+
+class CloudingEvent implements Comparable<CloudingEvent> {
+  CloudingEventType _type;
```

Comment:
it would be a bit cleaner if all those fields were final and we had an init constructor

Reply:
done